examples/acoustic/fwi_operators.py

```diff
@@ -20,66 +22,69 @@ class ForwardOperator(Operator):
-            4 * m * u + (s * damp - 2 * m) * u.backward +
-            2 * s**2 * (laplacian + s**2 / 12 * biharmonic))
-        # Add substitutions for spacing (temporal and spatial)
-        subs = {s: dt, h: model.get_spacing()}
-        # Receiver initialization
-        rec = SourceLike(name="rec", npoint=nrec, nt=nt, dt=dt, h=model.get_spacing(),
-                         coordinates=data.receiver_coords, ndim=len(damp.shape),
-                         dtype=damp.dtype, nbpml=model.nbpml)
-        source = SourceLike(name="src", npoint=nsrc, nt=nt, dt=dt, h=model.get_spacing(),
-                            coordinates=src.receiver_coords, ndim=len(damp.shape),
-                            dtype=damp.dtype, nbpml=model.nbpml)
-        source.data[:] = src.traces[:]
+    # Create the stencil by hand instead of calling numpy solve for speed purposes
+    # Simple linear solve of a u(t+dt) + b u(t) + c u(t-dt) = L for u(t+dt)
+    stencil = 1 / (2 * m + s * damp) * (
+        4 * m * u + (s * damp - 2 * m) * u.backward +
+        2 * s**2 * (laplacian + s**2 / 12 * biharmonic))
+    # Add substitutions for spacing (temporal and spatial)
+    subs = {s: dt, h: model.get_spacing()}
 
-        super(ForwardOperator, self).__init__(nt, m.shape,
-                                              stencils=Eq(u.forward, stencil),
-                                              subs=subs,
-                                              spc_border=max(spc_order, 2),
-                                              time_order=2,
-                                              forward=True,
-                                              dtype=m.dtype,
-                                              **kwargs)
+    if legacy:
+        op = Operator(nt, m.shape, stencils=Eq(u.forward, stencil), subs=subs,
+                      spc_border=max(spc_order / 2, 2), time_order=2, forward=True,
+                      dtype=m.dtype, **kwargs)
 
         # Insert source and receiver terms post-hoc
-        self.input_params += [source, source.coordinates, rec, rec.coordinates]
-        self.output_params += [rec]
-        self.propagator.time_loop_stencils_a = source.add(m, u) + rec.read(u)
-        self.propagator.add_devito_param(source)
-        self.propagator.add_devito_param(source.coordinates)
-        self.propagator.add_devito_param(rec)
-        self.propagator.add_devito_param(rec.coordinates)
+        op.input_params += [src, src.coordinates, rec, rec.coordinates]
+        op.output_params += [rec]
+        op.propagator.time_loop_stencils_a = src.add(m, u) + rec.read(u)
+        op.propagator.add_devito_param(src)
+        op.propagator.add_devito_param(src.coordinates)
+        op.propagator.add_devito_param(rec)
+        op.propagator.add_devito_param(rec.coordinates)
+
+    else:
+        # Create stencil expressions for operator, source and receivers
+        eqn = Eq(u.forward, stencil)
+        src_add = src.point2grid(u, m, u_t=t, p_t=time)
```

Comment:
Ok, so `src.point2grid` creates multiple expressions to perform interpolation from `src` to `u`. The time indices in those expressions differ between those expressions because `u` is buffered and thus requires one of the modulo-derived time indices (`t0`, `t1`, `t2`), while `src` actually has an unrolled `time` dimensions (formerly `i3`). Since the `StencilKernel` does no implicit time substitutions under the hood the idea is to explicitly use "real time" (`time`) and "modulo time" (`t`) according to how data is meant to e accessed.